ddsketch/ddsketch.py

```diff
@@ -3,167 +3,247 @@
-            self.store, self._count, self._sum, self._min, self._max
+        return (
+            f"store: {self.store}, negative_store: {self.negative_store}, "
+            f"zero_count: {self.zero_count}, count: {self.count}, "
+            f"sum: {self.sum}, min: {self.min}, max: {self.max}"
         )
 
     @property
     def name(self):
+        """str: name of the sketch"""
         return "DDSketch"
 
     @property
     def num_values(self):
-        return self._count
+        """int: number of values in the sketch"""
+        return self.count
 
     @property
     def avg(self):
-        return float(self._sum) / self._count
+        """float: exact avg of the values added to the sketch"""
+        return self.sum / self.count
 
     @property
     def sum(self):
+        """float: exact sum of the values added to the sketch"""
         return self._sum
 
-    def get_key(self, val):
-        if val < -self.min_value:
-            return -int(math.ceil(math.log(-val) / self.gamma_ln)) - self.offset
-        elif val > self.min_value:
-            return int(math.ceil(math.log(val) / self.gamma_ln)) + self.offset
-        else:
-            return 0
-
     def add(self, val):
         """Add a value to the sketch."""
-        key = self.get_key(val)
-        self.store.add(key)
+
+        if val > self._min_possible:
+            key = int(math.ceil(math.log(val) * self.multiplier))
+            self.store.add(key)
+        elif val < -self._min_possible:
+            key = int(math.ceil(math.log(-val) * self.multiplier))
+            self.negative_store.add(key)
+        else:
+            self.zero_count += 1
 
         # Keep track of summary stats
-        self._count += 1
+        self.count += 1
         self._sum += val
-        if val < self._min:
-            self._min = val
-        if val > self._max:
-            self._max = val
+        if val < self.min:
+            self.min = val
+        if val > self.max:
+            self.max = val
+
+    def get_quantile_value(self, quantile):
+        """the approximate value at the specified quantile
+
+        Args:
+            quantile (float): 0 <= q <=1
 
-    def quantile(self, q):
-        if q < 0 or q > 1 or self._count == 0:
+        Returns:
+            the value at the specified quantile or np.NaN if the sketch is empty
+        """
+        if quantile < 0 or quantile > 1 or self.count == 0:
             return np.NaN
-        if q == 0:
-            return self._min
-        if q == 1:
-            return self._max
-
-        rank = int(q * (self._count - 1) + 1)
-        key = self.store.key_at_rank(rank)
-        if key < 0:
-            key += self.offset
-            quantile = -2 * pow(self.gamma, -key) / (1 + self.gamma)
-        elif key > 0:
-            key -= self.offset
-            quantile = 2 * pow(self.gamma, key) / (1 + self.gamma)
+        if quantile == 0:
+            return self.min
+        if quantile == 1:
+            return self.max
```

Comment:
An issue with this special handling of `q == 0` and `q == 1` is that it makes the quantile function not necessarily increasing (e.g., we may have `get_quantile_value(0)` greater than `get_quantile_value(q)` with `q > 0`). See https://github.com/DataDog/sketches-go/pull/2#discussion_r222275306.